aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferUtility.java

```diff
@@ -597,6 +605,79 @@ public TransferObserver upload(String key, File file, ObjectMetadata metadata, C
         return upload(getDefaultBucketOrThrow(), key, file, metadata, cannedAcl, listener);
     }
 
+    /**
+     * Starts uploading the inputStream to the <b>default</b> bucket, using the given key.
+     *
+     * @param key           The key in the specified bucket by which to store the new object.
+     * @param inputStream   The input stream to upload.
+     */
+    public void uploadInputStream(String key, InputStream inputStream) {
+        try {
+            final File file = writeInputStreamToFile(inputStream);
+            TransferObserver observer = upload(key, file);
+
+            observer.setTransferListener(new TransferListener() {
```

Comment:
You probably have one more opportunity to clear up some duplicate code. If you create a private method that creates a TransferListener, I think you'd be in pretty good shape here. Something like this:

```
private TransferListener getListenerFor(File file) {
        return new TransferListener() {
                @Override
                public void onStateChanged(int id, TransferState state) {
                    switch (state) {
                        case COMPLETED:
                            file.delete();
                            break;
                        case FAILED:
                            throw new RuntimeException("Transfer failed.");
                    }
                }

                @Override
                public void onProgressChanged(int id, long bytesCurrent, long bytesTotal) {}

                @Override
                public void onError(int id, Exception ex) {
                    throw new RuntimeException(ex);
                }
            }
    }
```

Then the two call sites would look like:
```
observer.setTransferListener(getListenerFor(file));
```